csrc/ops/indexing.cpp

```diff
@@ -0,0 +1,268 @@
+  auto src_dom = TensorDomain::noReductions(src->getMaybeRFactorDomain());
+
+  TORCH_CHECK(!self_dom.empty(), "scatter can not be applied to 0d tensor.");
+  TORCH_CHECK(
+      self_dom.size() == idx_dom.size() && self_dom.size() == src_dom.size(),
+      "self, index and src tensor should all have the same number of dimensions in scatter like ops.");
+  if (dim < 0) {
+    dim += (int)self_dom.size();
+  }
+  TORCH_CHECK(
+      dim >= 0 && dim < (int)self_dom.size(),
+      "Scatter on invalid axis, received: ",
+      dim,
+      " however tensor view only has ",
+      self_dom.size(),
+      " non-reduction dims.");
+
+  // The shape of output tensor is same as self tensor.
+  std::vector<IterDomain*> out_domain;
+  for (const auto i : c10::irange(self_dom.size())) {
+    out_domain.push_back(
+        IterDomainBuilder(self_dom[i])
+            .iter_type(
+                self_dom[i]->getIterType() == IterType::Iteration
+                    ? IterType::GatherScatter
+                    : self_dom[i]->getIterType())
+            .build());
+  }
+
+  TensorView* out_tensor = IrBuilder::create<TensorView>(
+      IrBuilder::create<TensorDomain>(
+          out_domain, TensorDomain::getContiguityFilledWith(out_domain, true)),
+      self->getDataType().value());
+
+  IrBuilder::create<ScatterOp>(
+      type, out_tensor, self, dim, index, src, out_domain[dim]);
+  return out_tensor->as<TensorView>();
+}
+
+TensorView* scatter(
+    TensorView* self,
+    int dim,
+    TensorView* index,
+    TensorView* src) {
+  return scatterOp(ScatterOpType::Set, self, dim, index, src);
+}
+
+TensorView* take_along_axis(TensorView* inp, TensorView* index, int64_t dim) {
+  const auto inp_domain =
+      TensorDomain::noReductions(inp->getMaybeRFactorDomain());
+  const auto idx_domain =
+      TensorDomain::noReductions(index->getMaybeRFactorDomain());
+
+  TORCH_CHECK(
+      !inp_domain.empty(), "take_along_axis can not be applied to 0d tensor.");
+  TORCH_CHECK(
+      idx_domain.size() == inp_domain.size(),
+      "The input and index tensor must have the same dimensions for take_along_axis");
+
+  if (dim < 0) {
+    dim += (int)idx_domain.size();
+  }
+
+  TORCH_CHECK(
+      dim >= 0 && dim < (int)inp_domain.size(),
+      "take_along_axis on invalid axis, received: ",
+      dim,
+      " however tensor view only has ",
+      inp_domain.size(),
+      " non-reduction dims.");
+
+  std::vector<IterDomain*> out_domain(idx_domain.size());
+
+  for (const auto i : c10::irange(idx_domain.size())) {
+    auto inp_id = inp_domain.at(i);
+    auto idx_id = idx_domain.at(i);
+
+    // Both inp_id and idx_id can be a broadcast domain, however,
+    // since the extent of inp_id cannot be smaller than the extent of
+    // idx_id, it doesn't make sense to have a broadcast input id and
+    // a non-broadcast index id. Specifically, when the extent of the
+    // input ID is one, the index extent must also be one.
```

Comment:
Why? Doesn't numpy's `take_along_axis` support broadcasting from 1 to N?

Reply:
OK, you're right. numpy allows broadcast inputs:

```
>>> a = np.array([[10]])
>>> ai = np.array([[0], [0]])
>>> np.take_along_axis(a, ai, 1)
array([[10],
       [10]])
```

Reply:
I tried to support such a case here: https://github.com/NVIDIA/Fuser/pull/206/files#diff-30a0322eff3f4e5e4acc2bd48624c386db2cfe7434a6d3b8b49d0e5c2bf62143R233-R265

While I believe that should what we should do, the newly added test with a broadcast input, `TakeAlongBroadcastInput_CUDA`, segfaults. I suspect it has the same root cause as #93, so the test is currently disabled.

I'll put this RP on hold until #93 is fixed, but I don't expect much of this PR would change.